server/plugin.go

```diff
@@ -280,15 +282,76 @@ func (p *Plugin) deleteGitlabUserToken(userID string) error {
 func (p *Plugin) getGitlabUserInfoByMattermostID(userID string) (*gitlab.UserInfo, *APIErrorResponse) {
 	var userInfo gitlab.UserInfo
 
-	if infoBytes, err := p.API.KVGet(userID + GitlabUserInfoKey); err != nil || infoBytes == nil {
-		return nil, &APIErrorResponse{ID: APIErrorIDNotConnected, Message: "Must connect user account to GitLab first.", StatusCode: http.StatusBadRequest}
+	infoBytes, err := p.API.KVGet(userID + GitlabUserInfoKey)
+
+	if err != nil || infoBytes == nil {
+		gitlabTokenBytes, err := p.API.KVGet(userID + GitlabMigrationTokenKey)
+		if err != nil || gitlabTokenBytes == nil {
```

Comment:
I feel that this logic should go in the place where we attempt to fetch the token, rather than where we fetch the user info. The token is what's relevant in this migration, so we would want to "fall back" on using the token on first use if necessary.

Reply:
Not sure i understand, I'm doing the migration here because userinfo is the first thing we query, and we only deal with token later which is pretty tied to the decision made about how to solve the issue in this PR generally.

If the migration need to be done, the query from KV will fail. Unless I change it so the key is the same and conditionally unmarshal to a type with/without the token field, but that just seems much more complicated.

Reply:
I was saying that in the spot where we already using the mutex, we would do:

- Check if token exists in new KV location
- If not, check in old KV location. If it's in the old location, migrate to new location
- Use token whether we got it from old or new location

Is this how you understood my comment?

Reply:
I think so. But in order to get to that point where we are already using the mutex, we have loaded the userinfo in a prior step (which is not locked under mutex, and `p.getGitlabUserInfoByMattermostID` is called from lots of different places in code depending on what event is being handled). But if the migration needs to be done, we will fail to load the userinfo from the KV at that point, which is why I put the migration there.

To me, this relates to the decision of placing the token handling logic close to the gitlab client (inside the callback, which can then be locked under mutex if necessary), and allowing the userinfo to be stored/retrieved freely anywhere without mutex which is why we needed to separate these pieces of data into two KV entries. But in this case, a mutex is still needed for the migration just because we are also migrating the token.

Reply:
> But if the migration needs to be done, we will fail to load the userinfo from the KV at that point, which is why I put the migration there.

I guess I don't understand this part. Why will we fail to load the userinfo in that case?

Reply:
We will fail to load it because I changed the KV key and datatype that it is unmarshalled to. I see that I can try to do it as you're saying, where I guess I would keep the same KV key and just add a new one for the token, but I anticipate this might be messier in some ways. I can look into it.

Reply:
I suppose if this makes it possible to avoid using the mutex in a second place in the code, that is preferable.

To clarify, do you imagine just keeping the old KV key for userdata and creating a new one for the token only, or are we still migrating to two new keys and deleting the old one?

Reply:
Actually I don't think it will help with the mutex as needing a refresh and needing a migration are still completely separate conditions. Also if the old userdata + token are to be loaded without mutex lock and the migration is done only later when we attempt to fetch token from the new key, it's a possible issue that we are carrying the old token around that was loaded without mutex although I guess I'd re-load it with mutex when migrating as I'm doing in the migration now. But that's an example of how the current solution seems cleaner / safer to me: just immediately doing the migration.

Reply:
> I see that I can try to do it as you're saying, where I guess I would keep the same KV key and just add a new one for the token

Yeah I think we should do it this way. I don't think there's any reason to relocate the `UserInfo` data to a new key. It should stay in its current location, whereas the token is the exception to the rule, that it has to go somewhere else.

Also sorry for missing these posts. It's easier for me to stay up with conversations if I'm mentioned in the responses

> Actually I don't think it will help with the mutex as needing a refresh and needing a migration are still completely separate conditions.

Even though they're separate conditions, they touch the same data. I think they can be done in the same spot/mutex in the token usage block

> it's a possible issue that we are carrying the old token around that was loaded without mutex although I guess I'd re-load it with mutex when migrating as I'm doing in the migration now.  But that's an example of how the current solution seems cleaner / safer to me: just immediately doing the migration.

Yes I think we should re-load it with the mutex at the time of trying to use the token. I think having the migration at the spot of using the token would be cleaner, since we would just have one complicated spot where these store modifications will occur (migrate and token refresh)

Reply:
@hanzei What are you thoughts on the two strategies laid out above?

Reply:
@mickmister
- The old KV key for both token and userinfo is `_gitlabtoken`. If the token is no longer stored there but we keep the same key, it's a bit confusing.
- Before adding the migration, mutex was only used if we detect the token is expired. I don't see how we can add the migration functionality into the same function in order to minimize the use of the mutex. If it's something like "if token is expired, OR if migration needs to be performed, then run this function that both refreshes if necessary and migrates if necessary" seems like bad design, and that we should just have another function (as I currently do) that also uses the mutex. Or maybe use a wrapper like `.withTokenMutex` and then these two separate functions? That aside, I can still see the point in locating both mutex operations closer to each other.